python/dask_cudf/dask_cudf/core.py

```diff
@@ -304,6 +304,98 @@ def var(
                 result.divisions = (min(self.columns), max(self.columns))
             return handle_out(out, result)
 
+    def repartition_by_hash(
+        self,
+        columns=None,
+        npartitions=None,
+        max_branch=None,
+        disk_path=None,
+        ignore_index=True,
+        **kwargs,
+    ):
+        """Repartition a dask_cudf DataFrame by hashing.
+
+        Warning: By default, index will be ignored/dropped.
+
+        Parameter
+        ---------
+        columns : list, default None
+            List of columns (by name) to be used for hashing. If None,
+            all columns will be used.
+        npartitions : int, default None
+            Number of output partitions. If None, the output partitions
+            are chosen to match self.npartitions.
+        max_branch : int or False, default None
+            Passed to `rearrange_by_hash` - If False, single-stage shuffling
+            will be used (no matter the number of partitions).
+        disk_path : str, default None
+            If set to a string value, the repartitioning will be performed
+            "on disk," using a partitioned parquet dataset.
+        ignore_index : bool, default True
+            Ignore the index values while shuffling data into new
+            partitions. This can boost performance significantly.
+        kwargs : dict
+            Other `repartition` arguments.  Ignored.
+        """
+        npartitions = npartitions or self.npartitions
+        columns = columns or [col for col in self.columns]
+
+        # Use parquet-based shuffle on disk if path is provided
+        if isinstance(disk_path, str):
+            # WARNING: The `to_parquet` operation will use
+            # pyarrow until cudf#4236 (or similar) is merged
+            warnings.warn(
+                "Performing repartition_by_hash by writing a parquet"
+                " dataset - Expect poor performance!"
+            )
+            meta = self._meta._constructor_sliced([0])
+            partitions = self[columns].map_partitions(
+                batcher_sortnet.set_partitions_hash,
+                columns,
+                npartitions,
+                meta=meta,
+            )
+            df2 = self.assign(_partitions=partitions)
```

Comment:
`.assign` is a deep copy, is there any way we can use a shallow copy here?

Reply:
The actual application of `assign` should use a shallow copy in dask.dataframe (this was one of our recent optimizations).  However, I will double check this.